### PR TITLE
MRG+1: Add FIF montage reading

### LINKED ISSUE #2700: Incorrect EEG lead field matrices due to missing reference electrode location
The location of the EEG reference electrode is not stored in the raw data, leading to incorrect lead field matrices.

The EEG reference location is used by the forward solution in the BEM calculation, accounting for the fact that the reference is effectively subtracted out from the signal during acquisition. 

The forward solution code in both MNE C and MNE Python expects the reference electrode location to be stored in the second column of the position matrix (in python: info[chs][<#>][eeg_loc]). In the Python distribution, there is no way to populate this column.

For my data at least, the information is stored in the hpts montage file, so it would make sense to load the reference electrode location when the montage is loaded.

From the mne_analysis mailing list:
https://mail.nmr.mgh.harvard.edu/pipermail//mne_analysis/2015-December/003012.html

Code to replicate the mismatch:
https://drive.google.com/file/d/0B8PjVL8SDL7IdDVHWE40YTJSZnM/view?usp=drive_web

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2700.

cc @emilyps14.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
I'm working on getting `read_dig_montage` to do what I need it to do (read Neuromag dig data), and running into a few snags that I can't sort out from the `DigMontage` and `read_dig_montage` docstrings and code. cc @teonlamont @dengemann since it looks like you might know the code from `git blame`.

1. So far I think I've sorted out that `read_montage` and `read_dig_montage` are used for generic and specific/individualized digitization montages. I plan to update docstrings to reflect that.

2. The docstring for `read_dig_montage` makes it sound like `hsp` can (or is meant to?) hold the positions of channels in 3D. Is this also where the "extra" dig points go? If so, do the channels come first? If not, is there no way to save the "extra" points?

3. There is no way currently to associate the `hsp` points with particular channel names or types, right? Should there be? If `hsp` holds both channels and extra names, how is the delineation known?

4. What units are `hsp`, `hpi`, and `elp` supposed to be in when they're passed to the `DigMontage` constructor? It looks like with `read_dig_montage(..., transform=True)` that `elp` and `hsp` get transformed to device coordinates, while `hpi` does not because it's already in device space, and thus they should all be in device coords. This is consistent with the docstring of `read_dig_montage`, as `elp` is said to be in head space (in the file), which makes sense that `transform` would change it; `hpi` is already (on read) in device space, so it makes sense `transform` would not change it; and it is not specified which coordinate frame `hsp` belongs to, but can I safely infer it should be read in `head` coordinates? I'll update the docstring if so.

5. All of these points from `read_dig_montage` get passed to `DigMontage` constructor. Its docstring states that `hpi` are in device coords, doesn't say anything about the coord frame of `hsp`, and `elp` is said to be in "acquisition digitizer space". What does that mean? Whatever it means, it seems like we can't know from an instance of the `DigMontage` class what space each set of points is actually in, given that `read_dig_montage` could have points in either `head` or `device` coords depending on the `transform=True | False` input, and the state / coordinate frame is never saved. It seems like there is potential for error here. Is the idea that users should know whether or not they need to do a transform to get the points to the proper space (I guess that would be device for all points)? If so, it seems like we should update the `DigMontage` docstring to say that all points are assumed to be in Neuromag device coordinates.

6. I am confused about the `transform` vs `dev_head_t` arguments. It seems like both of them will construct a `device<->head` transform, just in different ways (one `head->dev`, the other `dev->head`). Why use a different method for constructing the transform, instead of constructing the `head_dev_t` once (based on the positions of the cardinal landmarks in head coord frame through `get_ras_to_neuromag_trans`) and inverting the result if you need `dev_head_t`?

7. If one does `read_dig_montage(..., dev_head_t=False)` (default), then eventually you get returned `DigMontage(..., dev_head_t=np.eye(4))`, even if you called `read_dig_montage` with `transform=True` (also the default). Note in this case, the inverse transform `head_dev_t` has already been constructed earlier, and is almost certainly not `np.eye(4)`. So in the resulting `DigMontage`, all the points will be in device coords, but the `dev_head_t` will be identity. It seems like this disobeys the definition of the head coordinate frame (as the origin being related to the positions of the LPA/RPA/Nasion) by making it equivalent to that of the device coord frame. So I guess my question is, why does the `DigMontage(..., dev_head_t)` only get constructed if `dev_head_t` is `True`?

I imagine these arguments might have been added incrementally to solve certain problems for specific use cases, but it's a bit challenging to sort out.

**Comment by teonlamont** [comment]
hey @Eric89GXL, will be answering in chunks.

the read_montage is where the channels go whereas read_dig_montage is where the headshape in head space, the marker in dev space (hpi), and digitized markers in head space (elp) go.

> The docstring for read_dig_montage makes it sound like hsp can (or is meant to?) hold the positions of channels in 3D.

these are the digitized headshape points that's why they are extra

**Comment by Eric89GXL** [comment]
> these are the digitized headshape points that's why they are extra

But in the code we have:

```
class DigMontage(object):
    """Montage for Digitized data

    Montages are typically loaded from a file using read_dig_montage. Only use
    this class directly if you're constructing a new montage.

    Parameters
    ----------
    hsp : array, shape (n_points, 3)
        The positions of the channels in 3d.
```
and
```
    hsp : None | str | array, shape (n_points, 3)
        If str, this corresponds to the filename of the headshape points.
        This is typically used with the Polhemus FastSCAN system.
        If numpy.array, this corresponds to an array of positions of the
        channels in 3d.
```
Is "channels" a typo in both places and it should say "extra points" instead, then?

**Comment by Eric89GXL** [comment]
> read_dig_montage is where the headshape in head space, the marker in dev space (hpi), and digitized markers in head space (elp) go.

Meaning those are the coordinate frames on disk, or that those are the coordinate frames they should be in by the time they are put in the `DigMontage` class?

By default they do all end up in device coords because `read_dig_montage(..., transform=True)` by default, but the also default kwarg `dev_head_t=False` means `DigMontage` thinks `dev_head_t=np.eye(4)`, which will be wrong. Do you agree?

Also, `transform` is an option, meaning that sometimes they might not be transformed at all -- is that because sometimes all three are already in device coords on disk? Or are `hsp` and `elp` sometimes meant to be kept in head coords (even though `DigMontage` doesn't store that information in any way)?

**Comment by teonlamont** [comment]
do you have any other questions about the DigMontage? sorry the docstrings were less than stellar

**Comment by agramfort** [comment]
let me know if you need to hangout.

@teonlamont thanks for helping eric on this.

it's a mess with emily currently.

thanks

File: mne/channels/montage.py [hunk 1]
```diff
@@ -129,7 +129,7 @@ def read_montage(kind, ch_names=None, path=None, unit='m', transform=False):
     kind : str
         The name of the montage file (e.g. kind='easycap-M10' for
         'easycap-M10.txt'). Files with extensions '.elc', '.txt', '.csd',
-        '.elp', '.hpts' or '.sfp' are supported.
+        '.elp', '.hpts', '.sfp', and '.fif' are supported.
```
Comment (jasmainak) [inline mne/channels/montage.py]: don't forget to update the documentation! It should go to `io.rst`

**Comment by jasmainak** [comment]
> 1. So far I think I've sorted out that read_montage and read_dig_montage are used for generic and specific/individualized digitization montages.

@Eric89GXL : from what I understand the user can also supply a montage file to `read_dig_montage`. You also need to give the `path` in that case.

File: mne/channels/montage.py (hunk 1 above)
Comment (emilyps14) [inline mne/channels/montage.py]: Sorry for making a mess guys :)

Comment (agramfort) [inline mne/channels/montage.py]: you point out our mess :)

File: mne/channels/montage.py [hunk 2]
```diff
@@ -259,9 +259,24 @@ def read_montage(kind, ch_names=None, path=None, unit='m', transform=False):
...
+        with f as fid:
+            dig = _read_dig_fif(fid, tree)
+        if dig is None:
+            raise RuntimeError('Digitizer data not found')
+        pos = []
+        ch_names_ = []
+        for d in dig:
+            if d['kind'] == FIFF.FIFFV_POINT_EEG:
+                pos.append(d['r'])
+                ch_names_.append('EEG%03.0f' % d['ident'])
```
Comment (teonlamont) [inline mne/channels/montage.py]: Do these channel names need a space between the type and number?

**Comment by dengemann** [comment]
I think it's not the first time I'm saying this but `read_dig_montage` is to me a misnomer and mildly confusing. I think the main problem is that as `read_montage` is for standard electrode positions, and not head shape points. I don't want to necessarily trigger the annoying deprecation game. We should have more clearly separated it in the first place. Of course it can get more confusing when digitization points happen to be aligned with EEG electrode positions, but still it is a different thing.

**Comment by agramfort** [comment]
+100

read_montage should be ONLY for standard montage. If you have dig points
you should use read_dig_montage.

that's the only clear way I see to explain it.

**Comment by Eric89GXL** [comment]
So should `read_dig_montage be changed to return the channel names and positions? It's easy to do for the FIF montage file I have but I have no idea about other systems.

**Comment by agramfort** [comment]
@teonlamont can you paste the code you write these days to use these
functions?

**Comment by Eric89GXL** [comment]
BTW @agramfort @dengemann `read_montage` already allows reading digitized channel positions by setting both the `name, ..., path` arguments. If I do `name=fname, path=''` for example it will read (channel) positions from a file I give it.

**Comment by agramfort** [comment]
yes but it's meant to be used for a standard montage that we may not ship.

**Comment by jasmainak** [comment]
> yes but it's meant to be used for a standard montage that we may not ship.

This is confusing and not apparent ... I am probably repeating myself in every PR but we need better documentation :\ I'm going to make another Documentation PR after the EEGLAB reader gets merged.

**Comment by teonlamont** [comment]
here's some pseudocode how we use it with the eeg digitizer data.

```python
raw = mne.io.read_raw_brainvision(filename)
digmontage = mne.read_dig_montage(hsp, hpi, elp, point_names, dev_head_t=True)
raw.set_montage(digmontage)
```

maybe @dengemann is right at the function name is confusing. The idea is to get digitizer data in for non-Elekta systems (since that's typically integrated in the `FIFF`). I modeled it to create an object that stored these data and can be added to the MNE object.

one solution would be to modify the DigMontage to store digitized channels (`DigMontage.dig_channels`)  then modify the `set_montage` function to apply them to the `info['chs']`.

**Comment by teonlamont** [comment]
@jasmainak, yeah, I agree. documentation is important. I will start writing up things that I've worked on that needs more explaining. this should go in the `doc/manual` right?

**Comment by jasmainak** [comment]
yep, that would be great @teonlamont :) If everyone contributes a little bit, documentation will look great in no time. Yes, I would put it in `doc/manual`.

**Comment by agramfort** [comment]
>
> here's some pseudocode how we use it with the eeg digitizer data.
>
> raw = mne.io.read_raw_brainvision(filename)
> digmontage = mne.read_dig_montage(hsp, hpi, elp, point_names, dev_head_t=True)
> raw.set_montage(digmontage)
>
> this looks clean. What I would like to avoid is having to manage both a
Montage and a DigMontage for the same subject.

I am fine having set_montage accepting either a Montage or a DigMontage.

> maybe @dengemann <https://github.com/dengemann> is right at the function
> name is confusing. The idea is to get digitizer data in for non-Elekta
> systems (since that's typically integrated in the FIFF). I modeled it to
> create an object that stored these data and can be added to the MNE object.
>
> one solution would be to modify the DigMontage to store digitized channels
> (DigMontage.dig_channels) then modify the set_montage function to apply
> them to the info['chs'].
>

ok so you need something extra after your set_montage?

**Comment by teonlamont** [comment]
> this looks clean. What I would like to avoid is having to manage both a
> Montage and a DigMontage for the same subject.

I tried to do that originally but I think @dengemann didn't like the idea to having digitizer data and channel data in the same object. IIRC he felt that montages were for predefined files of electrode data whereas the digitized locations and of dig data should be handled elsewhere.

I like the FIFF approach of having the two together and it makes things simpler IMO.

**Comment by agramfort** [comment]
>
> this looks clean. What I would like to avoid is having to manage both a
> Montage and a DigMontage for the same subject.
>
> I tried to do that originally but I think @dengemann
> <https://github.com/dengemann> didn't like the idea to having digitizer
> data and channel data in the same object. IIRC he felt that montages were
> for predefined files of electrode data whereas the digitized locations and
> of dig data should be handled elsewhere.
>
yes I agree with this vision.

but in your script I see only a DigMontage. So apparently you don's a
Montage also.

thanks for clarifying for me.

**Comment by Eric89GXL** [comment]
> but in your script I see only a DigMontage. So apparently you don's a
> Montage also.
>
> thanks for clarifying for me.

Where would digitized EEG locations come in? That would have to be Montage
/ read_montage currently because DigMontage won't have it. So are we adding
channel position reading to DigMontage?

**Comment by agramfort** [comment]
>
> Where would digitized EEG locations come in? That would have to be Montage
> / read_montage currently because DigMontage won't have it. So are we adding
> channel position reading to DigMontage?


yeah that sounds like the natural move.

@teonlamont wdyt? in what type of file do you store these locations?

what is unclear for me is that on a neuromag system the eeg electrode
locations are part of the digitization points....

**Comment by teonlamont** [comment]
I updated the pseudocode to show how you deal with both the montage and the digmontage. I find with adding the channels there but the called something like `digitized_ch_pos` or `dig_ch_pos` so it distinguishes it from the montage and denote the difference. then in the `set_montage` method, you could add an `if DigMontage.dig_ch_pos` for allowing the channel loc to the info

**Comment by agramfort** [comment]
in this 

```
raw = mne.io.read_raw_edf(filename)
# channel layout
montage = mne.read_montage('biosemi32')
raw.set_montage(montage)
# load digitizer data
digmontage = mne.channels.read_dig_montage(hsp, None, elp, point_names, transform=True)
raw.set_montage(digmontage)
```

how are the custom dig locations associated with the ch names of the standard montage?

**Comment by teonlamont** [comment]
currently they aren't. this is the current way to do it: add a standard electrode location array and digitized data. what I propose is that we make the following
- make `hsp` --> `dig`
  This will allow for formats like FIFF to only require this argument while Polhemus digitization can remain the same.
- add the FIFF dig reader to the `read_dig_montage` function
- within `_set_montage` [function](https://github.com/mne-tools/mne-python/blob/master/mne/channels/montage.py#L554), we could say the following:
```python
if montage.dig_ch_pos:
    for ch, dig_ch in zip(info['chs'], montage.dig_ch_pos):
        ch['loc'] = dig_ch
```
then for this case, you would then do the following:
```python
raw = mne.io.read_raw_fif('filename')
digmontage = mne.channels.read_dig_montage('digitizer filename', transform=True, dev_head_t=True)
raw.set_montage(digmontage)
```

**Comment by agramfort** [comment]
>
> currently they aren't. this is the current way to do it: add a standard
> electrode location array and digitized data.
>
so your electrodes locations are not informed by the dig points.... this is
not super satisfying.

> what I propose is that we make the following
>
>    - make hsp --> dig This will allow for formats like FIFF to only
>    require this argument while Polhemus digitization can remain the same.
>    - add the FIFF dig reader to the read_dig_montage function
>    - within _set_montage function
>    <https://github.com/mne-tools/mne-python/blob/master/mne/channels/montage.py#L554>,
>    we could say the following:
>
> if montage.dig_ch_pos:
>     for ch, dig_ch in zip(info['chs'], montage.dig_ch_pos):
>         ch['loc'] = dig_ch
>
> then for this case, you would then do the following:
>
> raw = mne.io.read_raw_fif('filename')
> digmontage = mne.channels.read_dig_montage('digitizer filename', transform=True, dev_head_t=True)
> raw.set_montage(digmontage)
>
>
that would be cleaner.

**Comment by Eric89GXL** [comment]
>> then for this case, you would then do the following:
>> raw = mne.io.read_raw_fif('filename')
>> digmontage = mne.channels.read_dig_montage('digitizer filename', transform=True, dev_head_t=True)
>> raw.set_montage(digmontage)

> that would be cleaner.

This sounds like the cleanest approach to me as well. I'll give implementing it a shot. I don't know what to do about the channel locations for anything other than FIF, so for now I'm going to implement it for that case. Then @teonlamont you can fix it for your use case if it makes sense. Sound good?

**Comment by teonlamont** [comment]
> so your electrodes locations are not informed by the dig points.... this is
> not super satisfying.

yeah, but that is the whole Montage class, it's standard template. The only one to get them into it now is using the `.hpts` format and defining your own.



> This sounds like the cleanest approach to me as well. I'll give implementing it a shot. I don't know what to do about the channel locations for anything other than FIF, so for now I'm going to implement it for that case. Then @teonlamont you can fix it for your use case if it makes sense. Sound good?

this sounds good.

**Comment by agramfort** [comment]
looks clean to me.

**Comment by agramfort** [comment]
do we still want this?

**Comment by Eric89GXL** [comment]
We still need something like this to solve the use case from the mailing list. I might have time to get it done properly today.

**Comment by Eric89GXL** [comment]
Okay this should implement what we discussed. @teonlamont could you have a look and see if you agree?

**Comment by Eric89GXL** [comment]
@agramfort feel free to have a look as well, I think it's what we discussed

**Comment by teonlamont** [comment]
@Eric89GXL, so we are considering the FIF montage to be a channel montage? does it have any headshape info or is it just the digitized positions?

**Comment by Eric89GXL** [comment]
has channel and headshape info

**Comment by teonlamont** [comment]
it looks good to me. maybe in a separate PR, it would be nice if we could write any generic `DigMontage` to disk via the FIF format. It seems like the best option for non-FIF originating files to be saved all together

**Comment by Eric89GXL** [comment]
Thanks for the review. It should be pretty easy to do once the `dig` field is populated properly, but +1 for doing that in a separate PR.

**Comment by Eric89GXL** [comment]
@agramfort feel free to have a look and merge if you're happy

**Comment by agramfort** [comment]
merged by rebase !

**Comment by agramfort** [comment]
thx @Eric89GXL

**Pull request merged**